src/lib/core/prbuf.c

```diff
@@ -0,0 +1,359 @@
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "prbuf.h"
+#include "bit/bit.h"
+
+/**
+ * Partitioned ring buffer. Each entry stores size before user data.
+ * So the typical buffer looks like:
+ * HEADER uint32 DATA uint32 DATA ...
+ *
+ * We have to store offsets to be able to restore buffer (including
+ * all metadata) from raw pointer. Otherwise it is impossible to point
+ * out where head/tail are located.
+ */
+struct PACKED prbuf_header {
+	/**
+	 * Buffer's data layout can be changed in the future, so for the sake
+	 * of proper recovery of the buffer we store its version.
+	 */
+	uint32_t version;
+	/** Total size of buffer (including header). */
+	uint32_t size;
+	/**
+	 * Offset of the oldest entry - it is the first candidate to be
+	 * overwritten. Note that in contrast to iterator/entry - this offset
+	 * is calculated to the first byte of entry (i.e. header containing
+	 * size of entry).
+	 */
+	uint32_t begin;
+	/** Offset of the next byte after the last (written) record. */
+	uint32_t end;
+};
+
+/**
+ * Structure representing record stored in the buffer so it has the same
+ * memory layout.
+ */
+struct PACKED prbuf_record {
+	/** Size of data. */
+	uint32_t size;
+	/** Data. */
+	char data[];
+};
+
+/**
+ * Current prbuf implementation version. Must be bumped in case storage
+ * format is changed.
+ */
+static const uint32_t prbuf_version = 0;
+
+/**
+ * prbuf is assumed to store all metadata in little-endian format. Beware
+ * when decoding its content. Size overhead per store operation is 4 bytes;
+ * moreover not the whole space of the buffer can be used since we do not
+ * wrap entry if it doesn't fit til the buffer's end.
+ *
+ * There are several assumptions regarding the buffer:
+ * - The end of the buffer (in the linear sense) contains "end mark";
+ * - The minimal size of the buffer is restricted;
+ * - Iteration direction - from the oldest entry to the newest.
+ */
+
+/** A mark of unused space in the buffer: trash is located after this point. */
+static const uint32_t prbuf_end_position = (uint32_t)(-1);
+
+/** Before storing a data in the buffer we place its size (i.e. header). */
+static const size_t record_size_overhead = sizeof(struct prbuf_record);
+
+/** Real size of allocation is (data size + record's header). */
+static uint32_t
+prbuf_record_alloc_size(size_t size)
+{
+	return size + offsetof(struct prbuf_record, data);
+}
+
+/** Returns pointer to the next byte after end of given buffer. */
+static char *
+prbuf_linear_end(struct prbuf *buf)
+{
+	return (char *)buf->header + buf->header->size;
+}
+
+/** Returns pointer to the first writable byte of given buffer. */
+static char *
+prbuf_linear_begin(struct prbuf *buf)
+{
+	return (char *)buf->header + sizeof(struct prbuf_header);
+}
+
+/** Returns pointer to the next byte after the last written record. */
+static char *
+prbuf_current_raw(struct prbuf *buf)
+{
+	return prbuf_linear_begin(buf) + buf->header->end;
```

Comment:
Nit: if you measure offset from the last byte of header (in my implementation I used offsets from the beginning of header) - you can add `char data[];` to prbuf_header and and use it instead of prbuf_linear_begin.

Reply:
I know but don't really like to introduce new members to header; moreover `linear_begin()` looks quite 'natural' to me. So if you don't mind I skip your comment.